src/Runner/Runner.php

```diff
@@ -89,26 +119,245 @@ public function __construct(
         $this->cacheManager = $cacheManager;
         $this->directory = $directory ?? new Directory('');
         $this->stopOnViolation = $stopOnViolation;
+        $this->parallelConfig = $parallelConfig ?? ParallelConfig::sequential();
+        $this->input = $input;
+        $this->configFile = $configFile;
     }
 
     /**
-     * @return array<string, array{appliedFixers: list<string>, diff: string}>
+     * @param iterable<\SplFileInfo> $fileIterator
+     */
+    public function setFileIterator(iterable $fileIterator): void
+    {
+        $this->fileIterator = $fileIterator;
+    }
+
+    /**
+     * @return _RunResult
      */
     public function fix(): array
+    {
+        return $this->parallelConfig->getMaxProcesses() > 1 && null !== $this->input
+            ? $this->fixParallel()
+            : $this->fixSequential();
+    }
+
+    /**
+     * Heavily inspired by {@see https://github.com/phpstan/phpstan-src/blob/9ce425bca5337039fb52c0acf96a20a2b8ace490/src/Parallel/ParallelAnalyser.php}.
+     *
+     * @return _RunResult
+     */
+    private function fixParallel(): array
```

Comment:
Before my changes the `fix()` method was similarly long. I've modified it to be a semaphore for sequential/parallel and also extracted `fixFile()`. I am not sure if I can reasonably split `fixParallel()` to "increase readability and reduce complexity", since most of the code there are React's callbacks that use some prepared, common variables. I've added comments there to describe what each part does, used meaningful naming for variables... IMHO it's readable enough 🤷‍♂️. Since it's only about cosmetics, maybe we can do it later to not block this PR unnecessarily?

Reply:
you say "cosmetics", I say "I failed to understand this method" 😓

Reply:
no worry, will put another hour to crack it

Reply:
understood it now 👍🏻